Source/JavaScriptCore/jit/RegisterSet.h

```diff
@@ -27,183 +27,500 @@
+                continue;
+            out.print(comma, reg);
+            if (m_bits.get(reg.index()) && (m_upperBits.get(reg.index()) || conservativeWidth(reg) != conservativeWidthWithoutVectors(reg)))
+                continue;
+
+            if (m_bits.get(reg.index()))
+                out.print("↓");
+            else
+                out.print("↑");
+        }
+        out.print("]");
+    }
+
+    inline constexpr bool operator==(const RegisterSet& other) const { return m_bits == other.m_bits && m_upperBits == other.m_upperBits; }
+    inline constexpr bool operator!=(const RegisterSet& other) const { return m_bits != other.m_bits || m_upperBits != other.m_upperBits; }
+
 private:
-    void setAny(Reg reg) { set(reg); }
-    void setAny(JSValueRegs regs) { set(regs); }
-    void setAny(const RegisterSet& set) { merge(set); }
-    void setMany() { }
-    template<typename RegType, typename... Regs>
-    void setMany(RegType reg, Regs... regs)
+    RegisterBitmap m_bits = { };
+    RegisterBitmap m_upperBits = { };
+};
+
+constexpr RegisterSetBuilder::RegisterSetBuilder(RegisterSet set)
+    : m_bits(set.m_bits)
+    , m_upperBits(set.m_upperBits)
+{ }
+
+constexpr RegisterSet RegisterSetBuilder::buildAndValidate() const
+{
+#if ASSERT_ENABLED
+    for (Reg reg = Reg::first(); reg <= Reg::last(); reg = reg.next()) {
+        if (!m_bits.get(reg.index()) && !m_upperBits.get(reg.index()))
+            continue;
+
+        ASSERT(!m_upperBits.get(reg.index()) || m_bits.get(reg.index()));
+    }
+#endif
+    return RegisterSet(*this);
+}
+
+constexpr RegisterSet RegisterSetBuilder::buildWithLowerBits() const
+{
+    return RegisterSet(*this);
+}
+
+constexpr size_t RegisterSetBuilder::numberOfSetRegisters() const { return buildAndValidate().numberOfSetRegisters(); }
+size_t RegisterSetBuilder::numberOfSetGPRs() const { return buildAndValidate().numberOfSetGPRs(); }
+size_t RegisterSetBuilder::numberOfSetFPRs() const { return buildAndValidate().numberOfSetFPRs(); }
+
+class ScalarRegisterSet final {
+public:
+    constexpr ScalarRegisterSet() { }
+
+    inline constexpr ScalarRegisterSet(const RegisterSet& registers)
+        : m_bits(registers.m_bits)
     {
-        setAny(reg);
-        setMany(regs...);
     }
 
-    // These offsets mirror the logic in Reg.h.
-    static constexpr unsigned gprOffset = 0;
-    static constexpr unsigned fprOffset = gprOffset + MacroAssembler::numGPRs;
-    
+    inline ScalarRegisterSet& operator=(const RegisterSet& r)
+    {
+        m_bits = r.m_bits;
+        return *this;
+    }
+
+    inline constexpr unsigned hash() const { return m_bits.hash(); }
+    inline constexpr bool operator==(const ScalarRegisterSet& other) const { return m_bits == other.m_bits; }
+    inline constexpr bool operator!=(const ScalarRegisterSet& other) const { return m_bits != other.m_bits; }
+
+    inline constexpr RegisterSet set() const WARN_UNUSED_RETURN
```

Comment:
I think `set` name sounds strange, and confusing given that this is ScalarRegisterSet (`set` sounds like a method to set something). I think it should be `toRegisterSet()`. Let's avoid short confusing name.